Switch default memory format of to (and similar) operators to Preserve

VitalyFedyunin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #31227 Fix `.to` operator to preserve strides when `self` should be returned
* #31226 Revert "Switch default memory format of to (and similar) operators to Preserve"
* #31139 [WIP] Fix cudnn channels_last descriptio problem
* #31131 Add memory_format support to `zeros`, `ones`, `full` (still need tests)
* #30089 Switch default memory format of clone operator to Preserve
* **#30088 Switch default memory format of to (and similar) operators to Preserve**
* #30087 Switch default memory format of _like operators to Preserve

Differential Revision: [D18624984](https://our.internmc.facebook.com/intern/diff/D18624984)

VitalyFedyunin committed:
Switch default memory format of to (and similar) operators to Preserve

[ghstack-poisoned]

VitalyFedyunin committed:
rebase to master + fix tests on "Switch default memory format of to (and similar) operators to Preserve"




[ghstack-poisoned]

VitalyFedyunin committed:
fix tests code on "Switch default memory format of to (and similar) operators to Preserve"


Differential Revision: [D18624984](https://our.internmc.facebook.com/intern/diff/D18624984)

[ghstack-poisoned]

VitalyFedyunin committed:
add experimental code to check model on "Switch default memory format of to (and similar) operators to Preserve"


Differential Revision: [D18624984](https://our.internmc.facebook.com/intern/diff/D18624984)

[ghstack-poisoned]

VitalyFedyunin committed:
experiments with tag on "Switch default memory format of to (and similar) operators to Preserve"


Differential Revision: [D18624984](https://our.internmc.facebook.com/intern/diff/D18624984)

[ghstack-poisoned]

VitalyFedyunin committed:
rebase to master on "Switch default memory format of to (and similar) operators to Preserve"


Differential Revision: [D18624984](https://our.internmc.facebook.com/intern/diff/D18624984)

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit 9493c0d:
* **1/1** broken upstream at merge base 56de885 (see [grid view](https://dr.pytorch.org/master-timeline.html?count=1&sha1=56de8853dad19a050bc9ce5f89c3f7572ca8a15c&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35))
    * You may want to rebase on the `viable/strict` branch (see its [recency history](https://dr.pytorch.org/master-viable-commits.html)):
        * If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
          ```
          git fetch viable/strict
          git rebase --onto viable/strict $(git merge-base origin/master HEAD)
          ```
        * If your commit is older than `viable/strict`:
          ```
          git fetch viable/strict
          git rebase viable/strict
          ```
* **0/1** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=9493c0ddac60a8228d3cf87d44cabb6cb956f310).

### 1 failure *not* recognized by patterns:
| Job | Step |
| --- | --- |
| ![CircleCI](https://avatars0.githubusercontent.com/ml/7?s=12) <sup>pytorch_windows_build</sup> | <sup>Build</sup> |


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

gchanan commented on aten/src/ATen/native/TensorConversions.cpp:
```diff
@@ -21,7 +21,7 @@ static inline Device ensure_has_index(Device device) {
 static inline Tensor to_impl(const Tensor& self, const TensorOptions& options, bool non_blocking, bool copy, c10::optional<c10::MemoryFormat> optional_memory_format) {
   auto memory_format =
-      optional_memory_format.value_or(MemoryFormat::Contiguous);
+      optional_memory_format.value_or(MemoryFormat::Preserve);
```
wasn't this broken before?  Because if I didn't pass in a memory format, memory_format would be set to contiguous.

But then suggest_memory_format wouldn't be contiguous, so I create a new tensor and copy.  But that doesn't seem like the correct semantics for the function.

VitalyFedyunin replied:
Indeed, looks like (in master) if we have channels last tensor and call .to to same dtype etc. we will get contiguous output instead of self. 

This PR will solve the issue.

gchanan replied:
is that behavior in 1.4?  Should we introduce a patch to fix it?

gchanan replied:
it would probably be good to fix that in a separate patch -- in case this gets reverted (happens with BC changes), we can have the bug fix in without this.

gchanan replied:
and seems worthy of a test :).

VitalyFedyunin committed:
Fix flake8 on "Switch default memory format of to (and similar) operators to Preserve"


Differential Revision: [D18624984](https://our.internmc.facebook.com/intern/diff/D18624984)

[ghstack-poisoned]

VitalyFedyunin committed:
Fix flake8 on "Switch default memory format of to (and similar) operators to Preserve"


Differential Revision: [D18624984](https://our.internmc.facebook.com/intern/diff/D18624984)

[ghstack-poisoned]
